### PR TITLE
Kxi 21090 propagate null support to remaining kdb types

### PR DESCRIPTION
**Pull request by vgrechin-kx**
Supporting null mapping of arrow types according to mockup test in q

https://kxl.atlassian.net/browse/KXI-21090

### CODE REVIEW

File: src/ArrayWriter.cpp [hunk 1]
```diff
@@ -434,29 +448,64 @@ template<>
 void PopulateBuilder<arrow::Type::BOOL>(shared_ptr<arrow::DataType> datatype, K k_array, arrow::ArrayBuilder* builder, TypeMappingOverride& type_overrides)
 {
   auto bool_builder = static_cast<arrow::BooleanBuilder*>(builder);
-  PARQUET_THROW_NOT_OK(bool_builder->AppendValues((uint8_t*)kG(k_array), k_array->n));
+  if( type_overrides.null_mapping.have_boolean ){
+    std::vector<bool> null_bitmap( k_array->n );
+    for( auto i = 0ll; i < k_array->n; ++i ){
+      null_bitmap[i] = type_overrides.null_mapping.boolean_null != kG( k_array )[i];
```
Comment (nmcdonnell-kx) [inline src/ArrayWriter.cpp:454]: `static_cast<bool>`?

File: src/ArrayWriter.cpp [hunk 2]
```diff
@@ -434,29 +448,64 @@ template<>
...
 
 template<>
 void PopulateBuilder<arrow::Type::UINT8>(shared_ptr<arrow::DataType> datatype, K k_array, arrow::ArrayBuilder* builder, TypeMappingOverride& type_overrides)
 {
   auto uint8_builder = static_cast<arrow::UInt8Builder*>(builder);
-  PARQUET_THROW_NOT_OK(uint8_builder->AppendValues((uint8_t*)kG(k_array), k_array->n));
+  if( type_overrides.null_mapping.have_uint8 ){
+    std::vector<bool> null_bitmap( k_array->n );
+    for( auto i = 0ll; i < k_array->n; ++i ){
+      null_bitmap[i] = type_overrides.null_mapping.uint8_null != kG( k_array )[i];
```
Comment (nmcdonnell-kx) [inline src/ArrayWriter.cpp:470]: `static_cast<uint8_t>`

File: src/ArrayWriter.cpp [hunk 3]
```diff
@@ -434,29 +448,64 @@ template<>
...
 template<>
 void PopulateBuilder<arrow::Type::UINT16>(shared_ptr<arrow::DataType> datatype, K k_array, arrow::ArrayBuilder* builder, TypeMappingOverride& type_overrides)
 {
   auto uint16_builder = static_cast<arrow::UInt16Builder*>(builder);
-  PARQUET_THROW_NOT_OK(uint16_builder->AppendValues((uint16_t*)kH(k_array), k_array->n));
-  arrow::Status s;
+  if( type_overrides.null_mapping.have_uint16 ){
+    std::vector<bool> null_bitmap( k_array->n );
+    for( auto i = 0ll; i < k_array->n; ++i ){
+      null_bitmap[i] = type_overrides.null_mapping.uint16_null != kH( k_array )[i];
```
Comment (nmcdonnell-kx) [inline src/ArrayWriter.cpp:502]: `static_cast<uint16_t>`

File: src/ArrayWriter.cpp [hunk 4]
```diff
@@ -502,36 +560,80 @@ template<>
...
 template<>
 void PopulateBuilder<arrow::Type::HALF_FLOAT>(shared_ptr<arrow::DataType> datatype, K k_array, arrow::ArrayBuilder* builder, TypeMappingOverride& type_overrides)
 {
-  arrow::HalfFloatType hft;
   auto hfl_builder = static_cast<arrow::HalfFloatBuilder*>(builder);
-  PARQUET_THROW_NOT_OK(hfl_builder->AppendValues((uint16_t*)kH(k_array), k_array->n));
+  if( type_overrides.null_mapping.have_float16 ){
+    std::vector<bool> null_bitmap( k_array->n );
+    for( auto i = 0ll; i < k_array->n; ++i ){
+      null_bitmap[i] = type_overrides.null_mapping.float16_null != kH( k_array )[i];
```
Comment (nmcdonnell-kx) [inline src/ArrayWriter.cpp:598]: `static_cast<uint16_t>`

File: src/KdbOptions.h [hunk 5]
```diff
@@ -57,10 +76,33 @@ namespace Options
   const std::string NULL_MAPPING = "NULL_MAPPING";
 
   // Null mapping options
+  const std::string NM_NA = "na";
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:79]: Don't need this

File: src/KdbOptions.h [hunk 6]
```diff
@@ -74,39 +116,135 @@ namespace Options
...
   struct NullMapping
   {
-      enum class Type: int{
-            INT_16
-          , INT_32
-          , STRING
-          , LARGE_STRING
-      };
-
+      bool have_na;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:122]: don't need this

File: src/KdbOptions.h [hunk 7]
```diff
@@ -74,39 +116,135 @@ namespace Options
...
+      bool have_time32;
+      bool have_time64;
+      bool have_decimal;
+      bool have_duration;
+      bool have_month_interval;
+      bool have_day_time_interval;
+
+      using Binary = std::basic_string<unsigned char>;
+
+      void* na_null = nullptr;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:152]: don't need this

File: src/KdbOptions.h [hunk 8]
```diff
@@ -74,39 +116,135 @@ namespace Options
...
+      int64_t timestamp_null;
+      int32_t time32_null;
+      int64_t time64_null;
+      double decimal_null;
+      int64_t duration_null;
+      int32_t month_interval_null;
+      int64_t day_time_interval_null;
+
+      template<arrow::Type::type TypeId = arrow::Type::NA>
+      inline auto GetOption() const { return std::make_pair( true, na_null );}
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:188]: don't need this

File: src/KdbOptions.h [hunk 9]
```diff
@@ -74,39 +116,135 @@ namespace Options
...
+} // namespace Options
 
 template<>
-inline const ETraits< Options::NullMapping::Type >::Names ETraits< Options::NullMapping::Type >::names {
-    { Options::NullMapping::Type::INT_16, Options::NM_INT_16 }
-  , { Options::NullMapping::Type::INT_32, Options::NM_INT_32 }
-  , { Options::NullMapping::Type::STRING, Options::NM_STRING }
-  , { Options::NullMapping::Type::LARGE_STRING, Options::NM_LARGE_STRING }
+inline const ETraits<arrow::Type::type>::Options ETraits<arrow::Type::type>::options{
+    { arrow::Type::NA, arrowkdb::Options::NM_NA }
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:221]: don't need this

File: src/KdbOptions.h [hunk 10]
```diff
@@ -310,8 +438,11 @@ class KdbOptions
     }
   }
 
-  template<Options::NullMapping::Type TypeId>
-  auto GetNullMappingOption( bool& result );
+  template<arrow::Type::type TypeId>
+  inline void HandleNullMapping( const std::string& key, K value );
+
+  template<arrow::Type::type TypeId = arrow::Type::NA>
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:444]: don't need this

File: src/KdbOptions.h [hunk 11]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+
+    throw KdbOptions::InvalidOption( message );
+}
+
 template<>
-inline auto KdbOptions::GetNullMappingOption<Options::NullMapping::Type::INT_16>( bool& result )
+inline void KdbOptions::HandleNullMapping<arrow::Type::BOOL>( const std::string& key, K value )
 {
-    result = null_mapping_options.have_int16;
+  switch( value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:489]: same thing
```
if (value->t == -KB || value->t == -KG)
```

File: src/KdbOptions.h [hunk 12]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::BINARY>( const std::string& key, K value )
+{
+  switch( value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:662]: again these are really the same `kG` is identical to `kC`

File: src/KdbOptions.h [hunk 13]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+    break;
+  default:
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::LARGE_BINARY>( const std::string& key, K value )
+{
+  switch( value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:679]: again these are really the same `kG` is identical to `kC`

File: src/KdbOptions.h [hunk 14]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::FIXED_SIZE_BINARY>( const std::string& key, K value )
+{
+  switch( value->t ){
+  case -UU:
+    null_mapping_options.fixed_binary_null.assign( &kU( value )->g[0], sizeof( U ) );
+    null_mapping_options.have_fixed_binary = true;
+    break;
+  case KG:
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:701]: again these are really the same `kG` is identical to `kC`

File: src/KdbOptions.h [hunk 15]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+    break;
+  default:
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::DATE32>( const std::string& key, K value )
+{
+  if( -KI == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:717]:
```suggestion
  if( -KD == value->t ){

```

File: src/KdbOptions.h [hunk 16]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::DATE64>( const std::string& key, K value )
+{
+  if( -KJ == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:729]:
```suggestion
  if( -KP == value->t ){

```

File: src/KdbOptions.h [hunk 17]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::TIMESTAMP>( const std::string& key, K value )
+{
+  if( -KJ == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:741]:
```suggestion
  if( -KP == value->t ){

```

File: src/KdbOptions.h [hunk 18]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::TIME32>( const std::string& key, K value )
+{
+  if( -KI == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:753]:
```suggestion
  if( -KT == value->t ){

```

File: src/KdbOptions.h [hunk 19]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::TIME64>( const std::string& key, K value )
+{
+  if( -KJ == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:765]:
```suggestion
  if( -KN == value->t ){

```

File: src/KdbOptions.h [hunk 20]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::DURATION>( const std::string& key, K value )
+{
+  if( -KJ == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:789]:
```suggestion
  if( -KN == value->t ){

```

File: src/KdbOptions.h [hunk 21]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::INTERVAL_MONTHS>( const std::string& key, K value )
+{
+  if( -KI == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:801]:
```suggestion
  if( -KM == value->t ){

```

File: src/KdbOptions.h [hunk 22]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  }
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::INTERVAL_DAY_TIME>( const std::string& key, K value )
+{
+  if( -KJ == value->t ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:813]:
```suggestion
  if( -KN == value->t ){

```

File: src/KdbOptions.h [hunk 23]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+    null_mapping_error( key, value );
+  }
+}
 
-    return null_mapping_options.int16_null;
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::UINT8>( const std::string& key, K value )
+{
+  if( -KG == value->t ){
+    null_mapping_options.uint8_null = value->g;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:507]: Is it worth having a static_cast here?

File: src/KdbOptions.h [hunk 24]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+    null_mapping_error( key, value );
+  }
+}
 
-    return null_mapping_options.int32_null;
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::UINT16>( const std::string& key, K value )
+{
+  if( -KH == value->t ){
+    null_mapping_options.uint16_null = value->h;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:531]: Is it worth having a static_cast here?

File: src/KdbOptions.h [hunk 25]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::UINT32>( const std::string& key, K value )
+{
+  if( -KI == value->t ){
+    null_mapping_options.uint32_null = value->i;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:555]: Is it worth having a static_cast here?

File: src/KdbOptions.h [hunk 26]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+  else{
+    null_mapping_error( key, value );
+  }
+}
+
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::UINT64>( const std::string& key, K value )
+{
+  if( -KJ == value->t ){
+    null_mapping_options.uint64_null = value->j;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:579]: Is it worth having a static_cast here?

File: src/KdbOptions.h [hunk 27]
```diff
@@ -341,38 +472,388 @@ class KdbOptions
...
+    null_mapping_error( key, value );
+  }
+}
 
-    return null_mapping_options.string_null;
+template<>
+inline void KdbOptions::HandleNullMapping<arrow::Type::HALF_FLOAT>( const std::string& key, K value )
+{
+  if( -KH == value->t ){
+    null_mapping_options.float16_null = value->h;
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:603]: Is it worth having a static_cast here?

File: src/ArrayWriter.cpp (hunk 2 above)
Comment (vgrechin-kx) [reply src/ArrayWriter.cpp:470]: nope - uint8_null is of basic_string<uint8_t> type

**Pull request merged**